### ISSUE TITLE
debugging html errors in question.html is overly difficult

### ISSUE THREAD

**Issue by cbm755** [PrairieLearn/PrairieLearn#5896]
I've often have a difficult time debugging my `question.html`.  Especially the interplay between malformed tags and `pl-*-panel`.  Here's a MWE:

```html
<pl-question-panel>
  <p>
    Prologue
  <p>
</pl-question-panel>

<p>
  And the answer is
  <pl-integer-input answers-name="value"></pl-integer-input>
</p>

<pl-question-panel>
  <p>
    Epilogue
 </p>
</pl-question-panel>
```

Noite that `<p>` instead of `</p>` after "Prologue".  Because of that the submissions panels appear empty.

- - - - 

Part of me thinks, "well garbage in, garbage out eh?"  but maybe we can do better:

  * part of the problem here is that browsers are very tolerant of such mistakes.
  * could PL check the validity of my `question.html` automatically?

There are also some things that I (as a user) could do:

  * I could run a stand alone command-line linter as part of my question dev
  * I could run a stand alone linter as part of my course CI/CD workflow on each MR.
  * I'm not aware of an OOTB linter but if anyone has one, let's all start using it!

**Comment by cbm755** [comment]
My notes on linters:

### tidy

I tried `tidy` but it is very unhappy with `<pl-question-panel>`'s and so on.  It does catch my `<p>Prologue<p>` bug but its kind of buried in the noise.  I didn't try to see how to give it an Allow List of extra non-standard HTML tags but perhaps this is possible.

### xmllint:

I tried `xmllint question.html`: this is quite nice but my understanding is our input is not valid xml (we have a forest not a tree).  If I wrap my entire MWE in say `<div>...</div>` then I get good results at detecting my `<p>` bug:
```
xmllint question.html  --noout
question.html:9: parser error : Opening and ending tag mismatch: p line 8 and pl-question-panel
</pl-question-panel>
                    ^
question.html:21: parser error : Opening and ending tag mismatch: p line 6 and div
</div>
      ^
question.html:22: parser error : Premature end of data in tag pl-question-panel line 5

^
```

I also tried `xmllint question.html --html --noout` but this also chokes on the custom tags.

Anyone else have tools to share?

**Comment by cbm755** [comment]
Maybe we *should* have a top-level root node in `question.html`?  I just tried making one up:
```
<pl-question-root>
  ...
</pl-question-root>
```
and PL itself seems tolerant of that.

And now my `xmllint question.html --noout` passes on a real-life non-trivial file (embedded Geogebra javascript, some moustache, LaTeX etc)...

**Comment by firasm** [comment]
Just to quickly add to this, it seems that the in-browser editor on prairielearn.com has some sort of error-checker. I'd be curious to know if there's a version of it that I can use in my code to automatically check my html files (outside the PL environment).

<img width="581" alt="Screen Shot 2022-06-11 at 10 33 12 PM" src="https://user-images.githubusercontent.com/2507459/173217207-3e835fec-98c2-4bf2-b609-1c98a87edbfd.png">

**Comment by firasm** [comment]
@cbm755 : `xmllint` seems promising! I'll check it out. It might even be worth it to write a short wrapper script that slurps in `question.html` and simply adds `<pl-question-root>` at the beginning, and `</pl-question-root>` at the end and then runs it through a script (to check existing questions).

**Comment by cbm755** [comment]
Yes that is exactly what I did in my `prairielinter` bash script.  I could post and we can make it better together.  But maybe others have other/better tooling?

**Comment by cbm755** [comment]
pushed code here: https://github.com/cbm755/prairielint

File issues with ideas.

**Comment by eliotwrobson** [comment]
Isn't this feature request essentially just this?: https://github.com/PrairieLearn/PrairieLearn/pull/1170

Since we want to deploy an experimental renderer soon, it might be nice to bundle it with these warnings.

**Comment by reteps** [comment]
Related -- we do some linting via #11837

**Comment by reteps** [comment]
This is now fixed with https://github.com/reteps/tree-sitter-htmlmustache. We don't integrate that into PL; but that remains to be seen as a direction we need. https://www.prairielearn.com/about/blog/stylish-mustaches
